src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4454,12 +4454,52 @@ internal SafeContext GetValEscape(BoundExpression expr, SafeContext scopeOfTheCo
                     return GetValEscapeOfObjectInitializer(initExpr, scopeOfTheContainingExpression);
 
                 case BoundKind.CollectionInitializerExpression:
-                    var colExpr = (BoundCollectionInitializerExpression)expr;
-                    return GetValEscape(colExpr.Initializers, scopeOfTheContainingExpression);
+                    {
+                        var colExpr = (BoundCollectionInitializerExpression)expr;
+
+                        //    var c = new C() { element };
+                        //
+                        // is equivalent to
+                        //
+                        //    var c = new C();
+                        //    c.Add(element);
+                        //
+                        // So we check arg mixing of `(receiverPlaceholder).Add(element)` calls,
+                        // and make the result "scoped" if any call could cause the elements to escape.
+
+                        using var _ = new PlaceholderRegion(this, [(colExpr.Placeholder, SafeContext.CallingMethod)]) { ForceRemoveOnDispose = true };
```

Comment:
Added `ForceRemoveOnDispose = true` because otherwise it can happen that the placeholder is added twice causing an assert to fail.

Reply:
This might need to be extended to also handle the assert in the Add part, like I'm doing in a similar PR: https://github.com/dotnet/roslyn/pull/76263/commits/b64ee878d237bcf28fad525a9d7c3b79eb737aca

But the bigger question is if this is a viable approach in general - using PlaceholderRegion temporarily during the check.

Reply:
It looks like adding a PlaceholderRegion temporarily here works, and it is unfortunate that PlaceholderRegions are typically added during the Visit methods and not removed. I'm curious though: was there a reason it was necessary to add the PlaceholderRegion temporarily in this particular case rather than adding it once (and not removing), in the calling `VisitObjectCreationExpressionBase()`?

Reply:
> was there a reason it was necessary to add the PlaceholderRegion temporarily in this particular case rather than adding it once (and not removing), in the calling `VisitObjectCreationExpressionBase()`?

Yes. When GetValEscape is called on a collection initializer, I need to decide whether it can be returnable or must be scoped to current block. So I temporarily mark the receiver with CallingMethod scope to see if that is ref safe - but it might not be ref safe (and depending on that, I determine the ValEscape of the collection initializer). Put another way, the receiver scope is just speculative and hence I don't think it should be set permanently in the visitor.